# Alpha Beta Pruning - Tic-Tac-Toe

At first I will be make the tic tac toe agent faster by adding alpha beta pruning to it and then compare its results (Time taken) with the plain minmax algorithm.

In [1]:
import math
import timeit

In [2]:
player = 'x'
opponent = 'o'

def isMovesLeft(board):
    for x in range(3):
        for y in range(3):
            if(board[x][y] == '_'):
                return True    
    return False

def evaluationFunction(board):
    for x in range(3):
        if(board[x][0] == board[x][1] and board[x][1] == board[x][2]):
            if(board[x][0] == player):
                return 10
            elif(board[x][0] == opponent):
                return -10
            
    for y in range(3):
        if(board[0][y] == board[1][y] and board[1][y] == board[2][y]):
            if(board[0][y] == player):
                return 10
            elif(board[0][y] == opponent):
                return -10
    
    if(board[0][0] == board[1][1] and board[1][1] == board[2][2]):
        if(board[0][0] == player):
            return 10
        elif(board[0][0] == opponent):
            return -10
    
    if(board[0][2] == board[1][1] and board[1][1] == board[2][0]):
        if(board[0][2] == player):
            return 10
        elif(board[0][2] == opponent):
            return -10
    
    return 0

def minmax(board, depth, isMax, alpha, beta):
    score = evaluationFunction(board)
    if(score == 10):
        return score - depth
    
    if(score == -10):
        return score + depth
    
    if(isMovesLeft(board) == False):
        return 0
    
    if(isMax):
        best = -math.inf
        for x in range(3):
            for y in range(3):
                if(board[x][y] == '_'):
                    board[x][y] = player
                    best = max(best, minmax(board, depth+1, not isMax, alpha, beta))
                    board[x][y] = '_'
                    alpha = max(alpha, best)
                    if alpha >= beta:
                        break
        return best
    else:
        best = math.inf
        for x in range(3):
            for y in range(3):
                if(board[x][y] == '_'):
                    board[x][y] = opponent
                    best = min(best, minmax(board, depth+1, not isMax, alpha, beta))
                    board[x][y] = '_'
                    beta = min(best, beta)
                    if(alpha >= beta):
                        break
        return best

def findBestMove(board):
    bestVal = -math.inf
    bestMove = (-1, -1)
    for x in range(3):
        for y in range(3):
            if(board[x][y] == '_'):
                board[x][y] = player
                moveVal = minmax(board, 0, False, -math.inf, math.inf)
                board[x][y] = '_'
                if(moveVal > bestVal):
                    bestVal = moveVal
                    bestMove = (x, y)
    
    print("Value of best move is:", bestVal)
    return bestMove

In [4]:
board = [
    [ 'x', '_', 'o' ],
    [ '_', 'o', '_' ],
    [ '_', 'x', '_' ]
]
 
bestMove = findBestMove(board)
 
print("The Optimal Move is :")
print("ROW:", bestMove[0], " COL:", bestMove[1])

Value of best move is: 8
The Optimal Move is :
ROW: 2  COL: 0


In [4]:
# Finding the time taken by Minmax with alpha beta pruning

mysetup = "import math"
  
mycode = ''' 
player = 'x'
opponent = 'o'

def isMovesLeft(board):
    for x in range(3):
        for y in range(3):
            if(board[x][y] == '_'):
                return True    
    return False

def evaluationFunction(board):
    for x in range(3):
        if(board[x][0] == board[x][1] and board[x][1] == board[x][2]):
            if(board[x][0] == player):
                return 10
            elif(board[x][0] == opponent):
                return -10
            
    for y in range(3):
        if(board[0][y] == board[1][y] and board[1][y] == board[2][y]):
            if(board[0][y] == player):
                return 10
            elif(board[0][y] == opponent):
                return -10
    
    if(board[0][0] == board[1][1] and board[1][1] == board[2][2]):
        if(board[0][0] == player):
            return 10
        elif(board[0][0] == opponent):
            return -10
    
    if(board[0][2] == board[1][1] and board[1][1] == board[2][0]):
        if(board[0][2] == player):
            return 10
        elif(board[0][2] == opponent):
            return -10
    
    return 0

def minmax(board, depth, isMax, alpha, beta):
    score = evaluationFunction(board)
    if(score == 10):
        return score - depth
    
    if(score == -10):
        return score + depth
    
    if(isMovesLeft(board) == False):
        return 0
    
    if(isMax):
        best = -math.inf
        for x in range(3):
            for y in range(3):
                if(board[x][y] == '_'):
                    board[x][y] = player
                    best = max(best, minmax(board, depth+1, not isMax, alpha, beta))
                    board[x][y] = '_'
                    alpha = max(alpha, best)
                    if alpha >= beta:
                        break
        return best
    else:
        best = math.inf
        for x in range(3):
            for y in range(3):
                if(board[x][y] == '_'):
                    board[x][y] = opponent
                    best = min(best, minmax(board, depth+1, not isMax, alpha, beta))
                    board[x][y] = '_'
                    beta = min(best, beta)
                    if(alpha >= beta):
                        break
        return best

def findBestMove(board):
    bestVal = -math.inf
    bestMove = (-1, -1)
    for x in range(3):
        for y in range(3):
            if(board[x][y] == '_'):
                board[x][y] = player
                moveVal = minmax(board, 0, False, -math.inf, math.inf)
                board[x][y] = '_'
                if(moveVal > bestVal):
                    bestVal = moveVal
                    bestMove = (x, y)
    
    print("Value of best move is:", bestVal)
    return bestMove

board = [
    [ 'x', '_', '_' ],
    [ '_', '_', '_' ],
    [ '_', '_', '_' ]
]
 
bestMove = findBestMove(board)
 
print("The Optimal Move is :")
print("ROW:", bestMove[0], " COL:", bestMove[1])
'''

print("Time taken to execute:", timeit.timeit(setup = mysetup, stmt = mycode, number = 1)) 

Value of best move is: 6
The Optimal Move is :
ROW: 0  COL: 1
Time taken to execute: 0.06583340000000248


In [5]:
# Simple minmax algorithm

mysetup = "import math"

mycode = '''
player = 'x'
opponent = 'o'

def isMovesLeft(board):
    for x in range(3):
        for y in range(3):
            if(board[x][y] == '_'):
                return True    
    return False

def evaluationFunction(board):
    for x in range(3):
        if(board[x][0] == board[x][1] and board[x][1] == board[x][2]):
            if(board[x][0] == player):
                return 10
            elif(board[x][0] == opponent):
                return -10
            
    for y in range(3):
        if(board[0][y] == board[1][y] and board[1][y] == board[2][y]):
            if(board[0][y] == player):
                return 10
            elif(board[0][y] == opponent):
                return -10
    
    if(board[0][0] == board[1][1] and board[1][1] == board[2][2]):
        if(board[0][0] == player):
            return 10
        elif(board[0][0] == opponent):
            return -10
    
    if(board[0][2] == board[1][1] and board[1][1] == board[2][0]):
        if(board[0][2] == player):
            return 10
        elif(board[0][2] == opponent):
            return -10
    
    return 0

def minmax(board, depth, isMax):
    score = evaluationFunction(board)
    if(score == 10):
        return score - depth
    
    if(score == -10):
        return score + depth
    
    if(isMovesLeft(board) == False):
        return 0
    
    if(isMax):
        best = -math.inf
        for x in range(3):
            for y in range(3):
                if(board[x][y] == '_'):
                    board[x][y] = player
                    best = max(best, minmax(board, depth+1, not isMax))
                    board[x][y] = '_'
        return best
    else:
        best = math.inf
        for x in range(3):
            for y in range(3):
                if(board[x][y] == '_'):
                    board[x][y] = opponent
                    best = min(best, minmax(board, depth+1, not isMax))
                    board[x][y] = '_'
        return best

def findBestMove(board):
    bestVal = -math.inf
    bestMove = (-1, -1)
    for x in range(3):
        for y in range(3):
            if(board[x][y] == '_'):
                board[x][y] = player
                moveVal = minmax(board, 0, False)
                board[x][y] = '_'
                if(moveVal > bestVal):
                    bestVal = moveVal
                    bestMove = (x, y)
    
    print("Value of best move is:", bestVal)
    return bestMove

board = [
    [ 'x', '_', '_' ],
    [ '_', '_', '_' ],
    [ '_', '_', '_' ]
]

bestMove = findBestMove(board)

print("The Optimal Move is :")
print("ROW:", bestMove[0], " COL:", bestMove[1])
'''

print("Time taken to execute:", timeit.timeit(setup = mysetup, stmt = mycode, number = 1))

Value of best move is: 6
The Optimal Move is :
ROW: 0  COL: 1
Time taken to execute: 0.22063420000000633


So we can see that minmax along with alpha beta pruning is much faster than simple minmax algorithm. Here in the inital state simple minmax takes 0.22 seconds to come up with a move by when we add alpha beta pruning to it, it will solve it in 0.065 seconds only.